File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int n = coins.size();
        vector<vector<int>> dp(n, vector<int>(amount+1,1e9));
        for(int i = 0;i<=amount;i++)
        dp[0][i] = i%coins[0] == 0? i/coins[0] : 1e9;

        for(int idx = 1;idx < n;idx++){
            for(int j=0;j <= amount;j++){
                int nt = dp[idx-1][j];
                int t =1e9;
                if(coins[idx] <= j)
                t = 1+ dp[idx][j-coins[idx]];

                dp[idx][j] = min(t,nt);
            }
        }
        return dp[n-1][amount] >= 1e9?-1:dp[n-1][amount];
    }
};

File: 0322-coin-change/0322-coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int n = coins.length;
        
        int dp[][] = new int[n][amount+1];

        for(int i[]: dp)Arrays.fill(i,(int)1e9);

        for(int tar=0;tar<=amount;tar++)
           dp[0][tar] = (tar % coins[0] == 0)? tar / coins[0] : (int)1e9;

        for(int i=1;i<n;i++){
            for(int tar =0;tar<=amount;tar++){
                int np = dp[i-1][tar];

                int p = (int)1e9;

                if(coins[i] <= tar)
                p =  1 + dp[i][tar - coins[i]];

                dp[i][tar] = Math.min(p , np);
            }
        }
        return dp[n-1][amount] >= (int)1e9 ? -1 : dp[n-1][amount];  
    }
}